src/main/java/ui/TextLineNumber.java

```diff
@@ -0,0 +1,292 @@
+            throws BadLocationException
+    {
+        Rectangle r = component.modelToView( rowStartOffset );
+        int lineHeight = fontMetrics.getHeight();
+        int y = r.y + r.height;
+        int descent = 0;
+
+        if (r.height == lineHeight)  // default font is being used
+        {
+            descent = fontMetrics.getDescent();
+        }
+        else  // We need to check all the attributes for font changes
+        {
+            if (fonts == null)
+                fonts = new HashMap<String, FontMetrics>();
+
+            Element root = component.getDocument().getDefaultRootElement();
+            int index = root.getElementIndex( rowStartOffset );
+            Element line = root.getElement( index );
+
+            for (int i = 0; i < line.getElementCount(); i++)
+            {
+                Element child = line.getElement(i);
+                AttributeSet as = child.getAttributes();
+                String fontFamily = (String)as.getAttribute(StyleConstants.FontFamily);
+                Integer fontSize = (Integer)as.getAttribute(StyleConstants.FontSize);
+                String key = fontFamily + fontSize;
+
+                FontMetrics fm = fonts.get( key );
+
+                if (fm == null)
+                {
+                    Font font = new Font(fontFamily, Font.PLAIN, fontSize);
+                    fm = component.getFontMetrics( font );
+                    fonts.put(key, fm);
+                }
+
+                descent = Math.max(descent, fm.getDescent());
+            }
+        }
+
+        return y - descent;
+    }
+
+    @Override
+    public void caretUpdate(CaretEvent e)
+    {
+
+        int caretPosition = component.getCaretPosition();
+        Element root = component.getDocument().getDefaultRootElement();
+        int currentLine = root.getElementIndex( caretPosition );
+
+        if (lastLine != currentLine)
+        {
+            repaint();
+            lastLine = currentLine;
+        }
+    }
+
+    @Override
+    public void changedUpdate(DocumentEvent e)
+    {
+        documentChanged();
+    }
+
+    @Override
+    public void insertUpdate(DocumentEvent e)
+    {
+        documentChanged();
+    }
+
+    @Override
+    public void removeUpdate(DocumentEvent e)
+    {
+        documentChanged();
+    }
+
+    private void documentChanged()
+    {
+
+        SwingUtilities.invokeLater(new Runnable()
+        {
+            @Override
+            public void run()
+            {
+                try
+                {
+                    int endPos = component.getDocument().getLength();
+                    Rectangle rect = component.modelToView(endPos);
+
+                    if (rect != null && rect.y != lastHeight)
+                    {
+                        setPreferredWidth();
+                        repaint();
+                        lastHeight = rect.y;
+                    }
+                }
+                catch (BadLocationException ex) { /* nothing to do */ }
```

Comment:
Mandatory:
For exception. use Log to record it and throw RuntimeException to let it fail. For how to use log. You can see some examples in other files